model/testresult/testresult.go

```diff
@@ -0,0 +1,81 @@
+package testresult
+
+import (
+	"github.com/evergreen-ci/evergreen/db"
+	"github.com/evergreen-ci/evergreen/db/bsonutil"
+	mgo "gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+const (
+	collection = "testresults"
+)
+
+// TestResult contains test data for a task.
+type TestResult struct {
+	ID        bson.ObjectId `bson:"_id" json:"_id"`
+	Status    string        `json:"status" bson:"status"`
+	TestFile  string        `json:"test_file" bson:"test_file"`
+	URL       string        `json:"url" bson:"url,omitempty"`
+	URLRaw    string        `json:"url_raw" bson:"url_raw,omitempty"`
+	LogID     string        `json:"log_id,omitempty" bson:"log_id,omitempty"`
+	LineNum   int           `json:"line_num,omitempty" bson:"line_num,omitempty"`
+	ExitCode  int           `json:"exit_code" bson:"exit_code"`
+	StartTime float64       `json:"start" bson:"start"`
+	EndTime   float64       `json:"end" bson:"end"`
+
+	// Together, TaskID and Execution identify the task which created this TestResult
+	TaskID    string `bson:"task_id" json:"task_id"`
+	Execution int    `bson:"task_execution" json:"task_execution"`
+
+	// LogRaw is not persisted to the database
+	LogRaw string `json:"log_raw" bson:"log_raw,omitempty"`
+}
+
+var (
+	// BSON fields for the task struct
+	testResultStatusKey    = bsonutil.MustHaveTag(TestResult{}, "Status")
+	testResultLineNumKey   = bsonutil.MustHaveTag(TestResult{}, "LineNum")
+	testResultTestFileKey  = bsonutil.MustHaveTag(TestResult{}, "TestFile")
+	testResultURLKey       = bsonutil.MustHaveTag(TestResult{}, "URL")
+	testResultLogIDKey     = bsonutil.MustHaveTag(TestResult{}, "LogID")
+	testResultURLRawKey    = bsonutil.MustHaveTag(TestResult{}, "URLRaw")
+	testResultExitCodeKey  = bsonutil.MustHaveTag(TestResult{}, "ExitCode")
+	testResultStartTimeKey = bsonutil.MustHaveTag(TestResult{}, "StartTime")
+	testResultEndTimeKey   = bsonutil.MustHaveTag(TestResult{}, "EndTime")
+	testResultTaskIDKey    = bsonutil.MustHaveTag(TestResult{}, "TaskID")
+	testResultExecutionKey = bsonutil.MustHaveTag(TestResult{}, "Execution")
+)
+
+// Insert writes a test result to the database.
+func (t *TestResult) Insert() error {
+	return db.Insert(collection, t)
+}
+
+// ByTaskIDAndExecution creates a query to return test results from the testresults collection for a given task.
+func ByTaskIDAndExecution(taskID string, execution int) db.Q {
```

Comment:
use a receiver here

Reply:
is a receiver necessary? I feel like there will be scenarios where you just have a task ID string + execution and want to form this query without creating a struct

Reply:
I think a receiver is necessary if you hide db.Q... 

check out the way the models are implemented in the sink package: https://github.com/evergreen-ci/sink/tree/master/model

I think the interfaces are still evolving a bit... [also the docs](https://godoc.org/github.com/evergreen-ci/sink/model)

Reply:
I have decided not to use a receiver at all here since I think
```
t := []TestResult{}
t.ByTaskIDAndExecution("foo", 0)
```
is a little awkward compared to
```
t := ByTaskIdAndExecution("foo", 0)
```

Reply:
sgtm - that's the awkwardness I was hoping to avoid